rust/arrow/src/csv/reader.rs

```diff
@@ -99,7 +99,27 @@ fn infer_field_schema(string: &str) -> DataType {
 /// If `max_read_records` is not set, the whole file is read to infer its schema.
 ///
 /// Return infered schema and number of records used for inference.
-fn infer_file_schema<R: Read + Seek>(
+pub fn infer_file_schema<R: Read + Seek>(
+    reader: &mut R,
+    delimiter: u8,
+    max_read_records: Option<usize>,
+    has_header: bool,
+) -> Result<(Schema, usize)> {
+    let (schema, records_count) =
+        infer_schema_from_reader(reader, delimiter, max_read_records, has_header)?;
+    // return the reader seek back to the start
```

Comment:
for any other reviewers, this seek back to the start simply keeps the existing behavior.

Reply:
also added a comment in the function docstring about this.

Reply:
While documenting this behavior, I think it's not the right thing to do to always rewind offset back to 0. Shouldn't we rewind to whatever offset it was at before we read any data to infer the schema?

Reply:
Seeking back to the original location is what *I* personally would expect, but I don't know the history / rationale for the current behavior so there might be a reason for it. It looks like it has been there since the [initial commit](https://github.com/apache/arrow/commit/ac45f3210a194049ef35f49847dbc4ff5e70d48f#diff-bee180737bbaeb6af694ff28ffe48f0426aa5555c0eda3bdd690636f91764551R178)  @nevi-me  perhaps do you remember if there was any rationale for this behavior?

Reply:
The idea in the initial commit was to return the file to its initial position before schema inference. That position was expected to always be the start, unless we want to allow someone to hand over a buffer that's in some other position.

If there's such a use-case, I think it's acceptable/fine. I was mainly avoiding the case where:
- user runs schema inference on inputs
- user tries to read the input with the inferred schema, getting no records, or missing records used in inferrence

I was still learning Rust when I wrote that code, so it's perhaps the case that the behaviour isn't what we'd like.

> Shouldn't we rewind to whatever offset it was at before we read any data to infer the schema?

@houqp if that position isn't 0, then I'd agree that we should rewind to it.

Reply:
sounds good, i will push a change this weekend for that :+1: